### PR TITLE
fix Issue 5305 - intrinsic functions have @safe stripped of them in r…

### PR DESCRIPTION
**Pull request by WalterBright**
…elease mode

Seems to work, adding a test case to ensure it.

### CODE REVIEW

**Comment by andralex** [comment]
Auto-merge toggled on

**Comment by JackStouffer** [comment]
I seem to recall @yebblies was adamant that Phobos shouldn't be imported in DMD tests

**Comment by yebblies** [comment]
Yes.

**Comment by yebblies** [comment]
Auto-merge toggled off

**Comment by WalterBright** [comment]
This is a compiler test, not a library test, and as such a test for it properly belongs in the compiler suite.

File: test/runnable/test5305.d [hunk 1]
```diff
@@ -0,0 +1,7 @@
+// https://issues.dlang.org/show_bug.cgi?id=5305
```
Comment (mathias-lang-sociomantic) [inline test/runnable/test5305.d]: Shouldn't `// REQUIRE_ARGS: -release` be used here ?

**Comment by mathias-lang-sociomantic** [comment]
Given `std.math` [just forward](https://github.com/dlang/phobos/blob/1bc9544fbf056b9fb7d77ad0da3cda43cf121c5d/std/math.d#L1506-L1512) to `core.math`, couldn't you provide a module that does the same ?

File: test/runnable/test5305.d (hunk 1 above)
Comment (WalterBright) [inline test/runnable/test5305.d]: I don't see why.

**Comment by WalterBright** [comment]
The intrinsic is directly associated, by the compiler, to std.math. Not something else. That's what needs to be tested.

@yebblies is generally right, because when porting dmd to a new platform it is very convenient to get it to pass the compiler tests before attempting to compile/debug Phobos. This is why this particular test is in its own file - it can be easily skipped when porting. This case is an exception.

**Comment by mathias-lang-sociomantic** [comment]
The issue title is confusing. I read it again and I'm going to close the issue as duplicate of [4541](https://issues.dlang.org/show_bug.cgi?id=4541).

For reference:
[You mentioned](https://issues.dlang.org/show_bug.cgi?id=5305#c2) that the issue was actually taking a pointer to an intrinsic, however the title was left unchanged. Out of the 2 solutions you mentioned, one of them was partially implemented by @yebblies in https://github.com/dlang/phobos/pull/3599 (with tests). Now the underlying issue is still present - taking the pointer of an intrinsic doesn't work -. It's just that the test code has shifted to:

```D
import core.math;
void map(real function(real) f) { }
void main() { map(&sqrt); }
````

So I suggest to close this P.R. as well.

**Comment by WalterBright** [comment]
The test case is to verify that taking the address of sqrt in std.math works. This test case is valid, and should be pulled to resolve the issue. 4541 is something else.

**Comment by mathias-lang-sociomantic** [comment]
@WalterBright : The intrinsics are defined [in core.math](https://github.com/dlang/druntime/blob/cf475a3a5016277f2a4b48dd70d11b5ddf90dda1/src/core/math.d#L88-L93).

In DMD we have a duplicate definition, [one for core.math](https://github.com/dlang/dmd/blob/e15c5fb3331ecbbbdf00e7f4e32872e5bde56bdd/src/builtin.d#L179) and [one for std.math](https://github.com/dlang/dmd/blob/e15c5fb3331ecbbbdf00e7f4e32872e5bde56bdd/src/builtin.d#L219).  The duplication was introduced in #20 to allow for this very change, but was obviously forgotten for a couple of years.

Thanks to the Phobos P.R. I mentioned earlier, a symbol with the same mangling as the `std.math` intrinsic [is now present in Phobos](https://github.com/dlang/phobos/blob/1bc9544fbf056b9fb7d77ad0da3cda43cf121c5d/std/math.d#L1506-L1512).
Just `nm` over your `libphobos`:

```
nm /usr/lib/x86_64-linux-gnu/libphobos2.a | grep D3std4math3sinFNaNbNiNfeZe
0000000000000000 T _D3std4math3sinFNaNbNiNfeZe
```

**Comment by andralex** [comment]
Yah, it's reasonable to import `std` here - it's the sync between the compiler and std that's being tested (rare).

**Comment by andralex** [comment]
Auto-merge toggled on

**Comment by mathias-lang-sociomantic** [comment]
> Yah, it's reasonable to import std here - it's the sync between the compiler and std that's being tested (rare).

Did you read the previous comments ?

**Comment by yebblies** [comment]
What the hell Andrei?

This is not a compiler issue - it is a phobos issue.  Phobos did not provide a body for the function, so its address couldn't be taken.  The argument that this should be in the dmd test suite and import phobos is a load of garbage.

**Comment by WalterBright** [comment]
Phobos and the compiler need to work together to make this work.

**Comment by yebblies** [comment]
You should have a look at what I changed to make this work.  Nothing to do with the compiler's handling of intrinsics.

**Comment by WalterBright** [comment]
The point of this test is to ensure that the fix, whatever it was, remains working.

**Comment by yebblies** [comment]
It is testing that a phobos function can be used in a certain way - the test belongs in phobos.  From the compiler's perspective this is just a normal function.

**Pull request merged**